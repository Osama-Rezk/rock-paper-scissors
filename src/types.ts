export enum OptionsEnum {
  Paper = "paper",
  Rock = "rock",
  Scissors = "scissors",
}

export type Option = {
  name: string;
  img?: string;
  value: OptionsEnum;
};

export enum RoundResult {
  Tie,
  PlayerOneWin,
  PlayerTwoWin,
}
export const resultsMapPlayerMode: Record<RoundResult, string> = {
  [RoundResult.Tie]: "Tie",
  [RoundResult.PlayerOneWin]: "You Win",
  [RoundResult.PlayerTwoWin]: "You Lose",
};

export const resultsMapAutomaticMode: Record<RoundResult, string> = {
  [RoundResult.Tie]: "Tie",
  [RoundResult.PlayerOneWin]: "Computer One Win",
  [RoundResult.PlayerTwoWin]: "Computer Two Win",
};
