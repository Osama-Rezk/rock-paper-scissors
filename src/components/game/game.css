.game-container {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  padding: 20px;
  gap: 20px;
}

.mode-container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.control-container {
  display: flex;
  gap: 30px;
}

.round-result {
  min-width: 150px;
}

.arena-container {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  gap: 30px;
}
