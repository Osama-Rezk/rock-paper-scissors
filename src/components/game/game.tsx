import { useEffect, useState } from "react";

import { OptionsEnum, Option, RoundResult } from "../../types";
import {
  getCurrentResultsDescription,
  getDescriptionMapByMode,
  getRandomItemFromArray,
  getRoundResults,
} from "../../utils";
import { Button } from "../button";
import { OptionsList } from "./options-list";
import { Player } from "./player";

import "./game.css";

export const ALL_OPTIONS: Option[] = [
  {
    name: "Paper",
    value: OptionsEnum.Paper,
  },
  {
    name: "Rock",
    value: OptionsEnum.Rock,
  },
  {
    name: "Scissors",
    value: OptionsEnum.Scissors,
  },
];

interface GameProps {
  winningScore: number;
}

export function Game({ winningScore }: GameProps) {
  const [playerOneOption, setPlayerOneOption] = useState<OptionsEnum | null>(
    null
  );
  const [playerTwoOption, setPlayerTwoOption] = useState<OptionsEnum | null>(
    null
  );

  const [playerOneScore, setPlayerOneScore] = useState<number>(0);
  const [playerTwoScore, setPlayerTwoScore] = useState<number>(0);

  const [isAutomatic, setIsAutomatic] = useState<boolean>(false);

  function play(selectedOption?: OptionsEnum) {
    const playerOneChoice =
      selectedOption || getRandomItemFromArray<Option>(ALL_OPTIONS).value;
    const playerTwoChoice = getRandomItemFromArray<Option>(ALL_OPTIONS).value;

    setPlayerOneOption(playerOneChoice);

    setPlayerTwoOption(playerTwoChoice);

    const roundResult = getRoundResults(playerOneChoice, playerTwoChoice);

    if (roundResult === RoundResult.Tie) return;

    if (roundResult === RoundResult.PlayerOneWin) {
      setPlayerOneScore((prevScore) => prevScore + 1);
    } else {
      setPlayerTwoScore((prevScore) => prevScore + 1);
    }
  }

  useEffect(() => {
    if (playerOneScore === winningScore || playerTwoScore === winningScore) {
      const map = getDescriptionMapByMode(isAutomatic);
      setTimeout(() => {
        alert(
          map[
            playerOneScore > playerTwoScore
              ? RoundResult.PlayerOneWin
              : RoundResult.PlayerTwoWin
          ]
        );
        reset();
      }, 10);
    }
  }, [playerOneScore, playerTwoScore, isAutomatic, winningScore]);

  function toggleMode() {
    reset();
    setIsAutomatic(!isAutomatic);
  }

  function reset() {
    setPlayerOneOption(null);
    setPlayerTwoOption(null);

    setPlayerOneScore(0);
    setPlayerTwoScore(0);
  }

  return (
    <div className="game-container">
      <h1>Rock Paper Scissors</h1>

      <div className="mode-container">
        <h1>{isAutomatic ? "Computer VS Computer" : "You Vs Computer"}</h1>
        <div className="control-container">
          <Button className="btn-primary" onClick={toggleMode}>
            Switch Modes
          </Button>
          <Button className="btn-danger" onClick={reset}>
            Reset
          </Button>
        </div>
      </div>

      <div className="arena-container">
        <Player
          data-testid="player-one"
          playerChoice={playerOneOption}
          score={playerOneScore}
        />

        <div className="round-result" data-testid="round-result">
          {playerOneOption && playerTwoOption ? (
            <h2>
              {getCurrentResultsDescription(
                playerOneOption,
                playerTwoOption,
                isAutomatic
              )}
            </h2>
          ) : (
            <h2>Game</h2>
          )}
        </div>
        <Player
          data-testid="player-two"
          playerChoice={playerTwoOption}
          score={playerTwoScore}
        />
      </div>

      {!isAutomatic && <OptionsList play={play} options={ALL_OPTIONS} />}

      {isAutomatic && (
        <Button className="btn-primary" onClick={() => play()}>
          run
        </Button>
      )}
    </div>
  );
}
