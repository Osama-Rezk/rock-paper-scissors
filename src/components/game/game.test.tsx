import { act, render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { OptionsEnum } from "../../types";
import { Game } from "./game";

//since computer choice is a random process so we must mock it
jest.mock("../../utils", () => ({
  ...jest.requireActual("../../utils"),
  getRandomItemFromArray: () => ({
    name: "Scissors",
    value: "scissors",
    img: "",
  }),
}));

jest.spyOn(window, "alert").mockImplementation(() => {});

const sleep = (t: number) =>
  new Promise((resolve) => {
    setTimeout(() => {
      resolve(0);
    }, t);
  });

describe("testing Game  Component", () => {
  it("Game elements Is Rendered Correctly", () => {
    render(<Game winningScore={3} />);

    expect(screen.getByTestId("options-list")).toBeInTheDocument();
    expect(screen.getByTestId("player-one")).toBeInTheDocument();
    expect(screen.getByTestId("player-two")).toBeInTheDocument();
  });

  it("User Can Switch the Mode", () => {
    render(<Game winningScore={3} />);

    expect(screen.getByText("You Vs Computer")).toBeInTheDocument();

    const switchButton = screen.getByRole("button", {
      name: /switch mode/i,
    });

    act(() => {
      switchButton.click();
    });

    expect(screen.getByText("Computer VS Computer")).toBeInTheDocument();
  });

  /*
   1. we make Player one choose rock By clicking on Rock option 
   2. We Mocked the computer choice so computer choice will be scissors
   3. Assert that player one choice are rendered correctly 
   4. Since player One will win so we will Assert that his score is 1
       and the computer Score is 0
   5. the round results should By "You win" (match snap shot)
   5. Reset the Game And make sure all score are zeros 
  
  */

  it("User Can Player And Reset Game", async () => {
    const user = userEvent.setup();

    render(<Game winningScore={3} />);

    const playOneElementContainer = within(screen.getByTestId("player-one"));

    const playTwoElementContainer = within(screen.getByTestId("player-two"));

    const optionButton = screen.getByRole("button", {
      name: /rock/i,
    });

    await user.click(optionButton);

    expect(
      playOneElementContainer.getByText(OptionsEnum.Rock)
    ).toBeInTheDocument();

    expect(playOneElementContainer.getByRole("heading")).toHaveTextContent("1");

    expect(
      playTwoElementContainer.getByText(OptionsEnum.Scissors)
    ).toBeInTheDocument();

    expect(playTwoElementContainer.getByRole("heading")).toHaveTextContent("0");

    expect(screen.getByTestId("round-result")).toMatchSnapshot();

    const resetButton = screen.getByRole("button", {
      name: /reset/i,
    });

    await user.click(resetButton);

    expect(screen.getAllByText("0")).toBeTruthy();
  });

  /*
  1. press the rock three times 
  2. assert that player one score is displayed as three
      and  player two score is zero 
  3. assert the alert() will be called after player one
     reach to WINNING_SCORE
  */
  it("Player one will one the game and show alert", async () => {
    const user = userEvent.setup();
    const WINNING_SCORE = 3;

    render(<Game winningScore={WINNING_SCORE} />);

    const playOneElementContainer = within(screen.getByTestId("player-one"));

    const playTwoElementContainer = within(screen.getByTestId("player-two"));

    const optionButton = screen.getByRole("button", {
      name: /rock/i,
    });

    for (let i = 0; i < WINNING_SCORE; i++) {
      await user.click(optionButton);
    }

    expect(playOneElementContainer.getByRole("heading")).toHaveTextContent("3");

    expect(playTwoElementContainer.getByRole("heading")).toHaveTextContent("0");

    await act(async () => {
      await sleep(100);
      expect(window.alert).toBeCalled();
    });
  });
});
