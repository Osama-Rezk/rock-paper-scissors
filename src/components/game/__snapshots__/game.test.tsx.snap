// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Game  Component User Can Player And Reset Game 1`] = `
<div
  class="round-result"
  data-testid="round-result"
>
  <h2>
    You Win
  </h2>
</div>
`;
